torch/utils/data/datapipes/utils/common.py

```diff
@@ -20,6 +21,86 @@
 ]
 
 
+def validate_input_col(fn: Callable, input_col: Optional[Union[int, tuple, list]]):
+    """
+    Checks that function used in a callable datapipe works with the input column
+
+    This simply ensures that the number of positional arguments matches the size
+    of the input column. The function must not contain any non-default
+    keyword-only arguments.
+
+    Examples:
+        >>> def f(a, b, *, c=1):
+        >>>     return a + b + c
+        >>> def f_def(a, b=1, *, c=1):
+        >>>     return a + b + c
+        >>> assert validate_input_col(f, [1, 2])
+        >>> assert validate_input_col(f_def, 1)
+        >>> assert validate_input_col(f_def, [1, 2])
+
+    Notes:
+        If the function contains variable positional (`inspect.VAR_POSITIONAL`) arguments,
+        for example, f(a, *args), the validator will accept any size of input column
+        greater than or equal to the number of positional arguments.
+        (in this case, 1).
+
+    Args:
+        fn: The function to check.
+        input_col: The input column to check.
+
+    Raises:
+        ValueError: If the function is not compatible with the input column.
+    """
+    sig = inspect.signature(fn)
```

Comment:
This breaks in case no signature can be extracted from the callable. For example:

```py
from torchdata.datapipes.iter import IterableWrapper

dp = IterableWrapper(range(3)).map(str)
```

```
ValueError: no signature found for builtin type <class 'str'>
```

Since we also can't use `lambda`'s or local functions, we are now forced to do something like:

```py
def to_str(obj):
    return str(obj)

dp = IterableWrapper(range(3)).map(to_str)
print(list(dp))
```

```
['0', '1', '2']
```

Without this check, the top datapipe would print the same. 

Can we maybe relax this check? For example, just return if we can't get a signature and let the user decipher the error message in case something goes wrong at runtime. Optionally, we could also warn in that case. Although we would also need a flag to turn the warning off. Otherwise, users might get warnings from builtin composed datapipes, e.g. torchvision datasets, without doing anything wrong or even the possibility to do something against it.

So far there is no case inside torchvision that needs this other than pytorch/vision#6076. Still, the behavior might be unexpected given that `Mapper` / `.map` accepts all callables.

Reply:
I agree with @pmeier that this may be too strict. It seems like `inspect.signature` can't handle built-in function or function implemented in C.

I will open a PR to skip over the check if the signature cannot be examined. Warnings seem to make it overly complicated IMO; I think if it does raise a warning the users should look into it. I am open to suggestions to make this better as well.